### PR TITLE
The zone introduction

### PR DESCRIPTION
**Pull request by StannisMod**
The great update that introduces zones, zone statuses and management.
- added `ZoneReq` to `WorkUnit` as new type of req
- added `Zone` to represent zone with status
- added `ZoneTransition` to save information about statuses' changes
- added `ZoneTimeline` to manage zone statuses through the time
- integrated zone accounting to all timelines as an additional constraint
- added zone operators to `Genetic`
- added `AccessCards e.g. ZoneTransitions` to the Gant chart visualization

Now we are ready to do zone planning.

### CODE REVIEW

File: tests/scheduler/genetic/fixtures.py [hunk 1]
```diff
@@ -48,15 +46,16 @@ def setup_toolbox(setup_default_schedules) -> tuple:
...
-                                               setup_worker_pool,
-                                               size_of_population,
-                                               mutate_order,
-                                               mutate_resources,
-                                               setup_default_schedules,
-                                               rand,
-                                               work_estimator=work_estimator,
-                                               landscape=setup_landscape_many_holders,
-                                               verbose=False)[0], resources_border,
+    return (create_toolbox(setup_wg,
```
Comment (Timotshak) [inline tests/scheduler/genetic/fixtures.py:49]: it schould be create_toolbox_and_mapping_objects
or
from sampo.scheduler.genetic.schedule_builder import create_toolbox_and_mapping_objects as create_toolbox

File: sampo/scheduler/generic.py [hunk 2]
```diff
@@ -77,6 +77,10 @@ def ft_getter(worker_team):
 
                 c_st, c_ft, _ = timeline.find_min_start_time_with_additional(node, workers, node2swork, spec, None,
                                                                              assigned_parent_time, work_estimator)
+                max_zone_time_after = timeline.zone_timeline.find_min_start_time(node.work_unit.zone_reqs, c_st, c_ft - c_st)
+                if c_st != max_zone_time_after:
+                    print(f'22222 Start time: {c_st}, zone time: {max_zone_time_after}')
```
Comment (Timotshak) [inline sampo/scheduler/generic.py:82]: is this print necessary?

File: sampo/scheduler/generic.py [hunk 3]
```diff
@@ -153,6 +157,11 @@ def build_scheduler(self,
                 start_time = assigned_parent_time
                 finish_time += start_time
 
+            max_zone_time_after = timeline.zone_timeline.find_min_start_time(node.work_unit.zone_reqs, start_time,
+                                                                             finish_time - start_time)
+            #if start_time != max_zone_time_after:
+            print(f'333333 Start time: {start_time}, zone time: {max_zone_time_after}, exec_time: {finish_time - start_time}')
```
Comment (Timotshak) [inline sampo/scheduler/generic.py:163]: is this print necessary?

File: sampo/scheduler/timeline/just_in_time_timeline.py [hunk 4]
```diff
@@ -85,13 +93,37 @@ def find_min_start_time_with_additional(self, node: GraphNode,
...
 
-        c_st = max(c_st, max_material_time)
+        max_zone_time = self.zone_timeline.find_min_start_time(node.work_unit.zone_reqs, c_st, exec_time)
 
-        c_ft = c_st + calculate_working_time_cascade(node, worker_team, work_estimator)
+        c_st = max(c_st, max_material_time, max_zone_time)
+
+        max_zone_time_new = self.zone_timeline.find_min_start_time(node.work_unit.zone_reqs, c_st, exec_time)
+        if max_zone_time_new != c_st:
+            print('ERROR!!!')
```
Comment (Timotshak) [inline sampo/scheduler/timeline/just_in_time_timeline.py:124]: if we have an error, should we just continue the execution?

File: sampo/scheduler/timeline/momentum_timeline.py [hunk 5]
```diff
@@ -123,21 +124,66 @@ def apply_time_spec(time: Time):
...
+                                                                                   node.work_unit.need_materials(),
+                                                                                   node.work_unit.workground_size)
+                max_zone_time = self.zone_timeline.find_min_start_time(node.work_unit.zone_reqs, start_time, exec_time)
+
+                prev_st = st
+                st = max(max_material_time, max_zone_time, start_time)
+
+                max_zone_time_after = self.zone_timeline.find_min_start_time(node.work_unit.zone_reqs, st, exec_time)
+                if st != max_zone_time_after:
+                    print(f'2 Start time: {st}, zone time: {max_zone_time_after}')
```
Comment (Timotshak) [inline sampo/scheduler/timeline/momentum_timeline.py:179]: is this print necessary?

File: sampo/scheduler/timeline/momentum_timeline.py [hunk 6]
```diff
@@ -123,21 +124,66 @@ def apply_time_spec(time: Time):
...
+
+                assert st >= max_parent_time
+
+                if st.is_inf():
```
Comment (Timotshak) [inline sampo/scheduler/timeline/momentum_timeline.py:183]: This condition looks strange, is it really necessary?

File: sampo/scheduler/timeline/zone_timeline.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+        type2status: dict[str, int] = build_index(zones, lambda w: w.kind, lambda w: w.required_status)
+
+        queue = deque(zones)
+
+        i = 0
+        while len(queue) > 0:
+            # if i > 0 and i % 50 == 0:
+            #     print(f'Warning! Probably cycle in looking for time slot for all reqs: {i} iteration')
+            #     print(f'Current queue size: {len(queue)}')
```
Comment (Timotshak) [inline sampo/scheduler/timeline/zone_timeline.py:50]: rudiment comments

File: sampo/scheduler/timeline/zone_timeline.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            #     next_cpkt_time = state[next_cpkt_idx].time
+            #     if (parent_time <= next_cpkt_time <= start_time_changed
+            #             and self._match_status(state[next_cpkt_idx].available_workers_count, required_status)):
+            #         # waiting until the next checkpoint is faster that change zone status
+            #         current_start_time = next_cpkt_time
+            #         current_start_idx += 1
+            #     else:
+            #         current_start_time = start_time_changed
+            #     # renewing the end index
+            #     end_idx = state.bisect_right(current_start_time + exec_time)
```
Comment (Timotshak) [inline sampo/scheduler/timeline/zone_timeline.py:177]: rudiment comments

File: sampo/scheduler/timeline/zone_timeline.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+            if not not_compatible_status_found:
+                break
+
+            if current_start_idx >= len(state):
+                # cur_cpkt = state[-1]
+                # if cur_cpkt.time == current_start_time and not self._match_status(cur_cpkt.available_workers_count,
+                #                                                                   required_status):
+                #     # print('Problem!')
+                #     current_start_time = max(parent_time, state[-1].time + 1)
```
Comment (Timotshak) [inline sampo/scheduler/timeline/zone_timeline.py:203]: rudiment comments

File: sampo/scheduler/timeline/zone_timeline.py (hunk 7 above)
Comment (StannisMod) [reply sampo/scheduler/timeline/zone_timeline.py:50]: These lines should be uncommented when performance issues rise

File: sampo/schemas/scheduled_work.py [hunk 10]
```diff
@@ -127,8 +134,8 @@ def to_dict(self) -> dict[str, Any]:
...
         }
 
-    def __deepcopy__(self, memodict={}):
-        return ScheduledWork(deepcopy(self.work_unit, memodict),
-                             deepcopy(self.start_end_time, memodict),
-                             deepcopy(self.workers, memodict),
-                             self.contractor)
+    # def __deepcopy__(self, memodict={}):
+    #     return ScheduledWork(deepcopy(self.work_unit, memodict),
+    #                          deepcopy(self.start_end_time, memodict),
```
Comment (vanoha) [inline sampo/schemas/scheduled_work.py:139]: rudiment comments

**Review by vanoha — COMMENTED** [review]
Request changes

File: sampo/scheduler/timeline/zone_timeline.py (hunk 8 above)
Comment (StannisMod) [reply sampo/scheduler/timeline/zone_timeline.py:177]: this is the code that should be tested and added later

**Review by vanoha — APPROVED** [review]
Approved

**Pull request merged**